src/components/side-panel/layer-panel/layer-list.js

```diff
@@ -0,0 +1,174 @@
+const SortableStyledItem = styled.div`
+  z-index: ${props => props.theme.dropdownWrapperZ + 1};
+
+  &.sorting {
+    pointer-events: none;
+  }
+
+  &.sorting-layers .layer-panel__header {
+    background-color: ${props => props.theme.panelBackgroundHover};
+    font-family: ${props => props.theme.fontFamily};
+    font-weight: ${props => props.theme.fontWeight};
+    font-size: ${props => props.theme.fontSize};
+    line-height: ${props => props.theme.lineHeight};
+    *,
+    *:before,
+    *:after {
+      box-sizing: border-box;
+    }
+    .layer__drag-handle {
+      opacity: 1;
+      color: ${props => props.theme.textColorHl};
+    }
+  }
+`;
+
+LayerListFactory.deps = [LayerPanelFactory];
+
+function LayerListFactory(LayerPanel) {
+  // By wrapping layer panel using a sortable element we don't have to implement the drag and drop logic into the panel itself;
+  // Developers can provide any layer panel implementation and it will still be sortable
+  const SortableItem = SortableElement(({children, isSorting}) => {
+    return (
+      <SortableStyledItem className={classnames('sortable-layer-items', {sorting: isSorting})}>
+        {children}
+      </SortableStyledItem>
+    );
+  });
+
+  const WrappedSortableContainer = SortableContainer(({children}) => {
+    return <div>{children}</div>;
+  });
+
+  const LayerList = props => {
+    const {
+      layers,
+      datasets,
+      layerOrder,
+      uiStateActions,
+      visStateActions,
+      layerClasses,
+      isSortable
+    } = props;
+    const {toggleModal: openModal} = uiStateActions;
+    const [isSorting, setIsSorting] = useState(false);
+
+    const layerClassSelector = () => layerClasses;
+    const layerTypeOptionsSelector = createSelector(layerClassSelector, layerCls =>
+      Object.keys(layerCls).map(key => {
+        const layer = new layerCls[key]();
+        return {
+          id: key,
+          label: layer.name,
+          icon: layer.layerIcon,
+          requireData: layer.requireData
+        };
+      })
+    );
+
+    const layerTypeOptions = layerTypeOptionsSelector(props);
+
+    const layerActions = {
+      layerColorUIChange: visStateActions.layerColorUIChange,
+      layerConfigChange: visStateActions.layerConfigChange,
+      layerVisualChannelConfigChange: visStateActions.layerVisualChannelConfigChange,
+      layerTypeChange: visStateActions.layerTypeChange,
+      layerVisConfigChange: visStateActions.layerVisConfigChange,
+      layerTextLabelChange: visStateActions.layerTextLabelChange,
+      removeLayer: visStateActions.removeLayer,
+      duplicateLayer: visStateActions.duplicateLayer
+    };
+
+    const panelProps = {
+      datasets,
+      openModal,
+      layerTypeOptions
+    };
+
+    const _handleSort = ({oldIndex, newIndex}) => {
+      visStateActions.reorderLayer(arrayMove(props.layerOrder, oldIndex, newIndex));
+      setIsSorting(false);
+    };
+
+    const _onSortStart = () => {
```

Comment:
use `useCallback`

Reply:
This can't be added to useCallback since it doesn't have sideeffects. At least that's what VSC is telling me.